### PR TITLE
Introduce `errstate` configuration to control cuSOLVER `devInfo` and cuBLAS `infoArray` checks

### PR DESCRIPTION
**Pull request by hvy**
Fixes https://github.com/cupy/cupy/issues/2414.

**_Please review the design._**

This PR introduces an environment variable to control (turn on/off) the `devInfo` check after cuSOLVER routine calls. After discussions in the above issue, the default behavior was decided to not check the `devInfo`. This PR is labaled as `no-compat` since it will silence errors in existing code and break NumPy compatibility.

I'm open for discussion about the design and will continue adding tests and documentation once that looks okay.

**TODO:**
- [x] Implement
- [x] Test
- [x] Document

### CODE REVIEW

**Comment by hvy** [comment]
There are still opinions about unifying the default behavior to actually sync and check the devInfo.  Would be nice to agree on either before merging new PRs (e.g. https://github.com/cupy/cupy/pull/2495).

**Comment by hvy** [comment]
Added tests for routines where I could come up with negative cases (e.g. singular inputs). I'm not sure how to test the rest of them since the expected negative results are not necessarily obvious nor consistent with NumPy. PTAL.

**Comment by asi1024** [comment]
Could you fix to use `errstate` introduced in #2535?

File: cupy/linalg/norms.py [hunk 1]
```diff
@@ -240,17 +240,20 @@ def _slogdet_one(a):
     buffersize = getrf_bufferSize(handle, m, m, a_copy.data.ptr, m)
     workspace = cupy.empty(buffersize, dtype=dtype)
     getrf(handle, m, m, a_copy.data.ptr, m, workspace.data.ptr,
-          ipiv.data.ptr, info.data.ptr)
+          ipiv.data.ptr, dev_info.data.ptr)
+
+    try:
+        cupy.linalg.util._check_cusolver_dev_info_if_synchronization_allowed(
+            getrf, dev_info)
```
Comment (hvy) [inline cupy/linalg/norms.py]: Open for discussion. C.f. https://github.com/cupy/cupy/pull/2437/files#diff-2ef2aeb8e00bf17e47e54407373d8c27R163-R164.

File: cupyx/_ufunc_config.py [hunk 2]
```diff
@@ -1,28 +1,32 @@
...
+# In favor of performance, the `devInfo` input/output from cuSOLVER routine
+# calls that is necessary to check the validity of the other outputs, are
+# ignored, as D2H copy incurring device synchronizations would otherwise be
+# required.
+config.linalg = 'ignore'
 
 
 @contextlib.contextmanager
-def errstate(*, divide=None, over=None, under=None, invalid=None):
+def errstate(*, divide=None, over=None, under=None, invalid=None, linalg=None):
```
Comment (hvy) [inline cupyx/_ufunc_config.py]: I guess we should document all these arguments in all these functions, in a separate PR.

**Comment by hvy** [comment]
Updated to use `errstate`. PTAL.

**Comment by hvy** [comment]
Fixed `_batched_inv` as well (which actually calls a cuBLAS routine and not a cuSOLVER one). PTAl.

**Comment by hvy** [comment]
Let me update the title of this PR reflect the new changes.

File: cupy/linalg/solve.py [hunk 3]
```diff
@@ -344,20 +375,8 @@ def _batched_inv(a):
...
-        info = info_array[i]
-        if info < 0:
-            err = True
-            err_detail += ('\tmatrix[{}]: illegal value at {}-the parameter.'
-                           '\n'.format(i, -info))
-        if info > 0:
-            err = True
-            err_detail += '\tmatrix[{}]: matrix is singular.\n'.format(i)
-    if err:
-        raise RuntimeError('matrix inversion failed at getrf.\n' + err_detail)
```
Comment (hvy) [inline cupy/linalg/solve.py]: Note that this PR changes this error type to `numpy.linalg.LinAlgError`.

**Comment by asi1024** [comment]
LGTM. Jenkins, test this please.

**Comment by asi1024** [comment]
https://github.com/chainer/chainer/blob/59df4d08a0592cb95325020ab7d83e8b6bcd7079/tests/chainer_tests/functions_tests/math_tests/test_det.py#L208-L225

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**